[EarlyCSE] Rematerialize alignment assumption.

fhahn opened the pull request:
If the load has align metadata, preserve it via an alignment
assumption. Note that this doesn't use salavageKnowledge, as we need
to create the assumption for the value we replaced the load with.

Comparison run https://github.com/dtcxzyw/llvm-opt-benchmark/pull/1343#issuecomment-2361976028

Unfortunately the diff is too large to display, top changes
```
Top 5 improvements:
  typst-rs/4p30esqzpn2o5olu.ll 6364278007 6347884002 -0.26%
  llvm/StdLibraryFunctionsChecker.cpp.ll 25872800221 25821679537 -0.20%
  meshlab/arap.cpp.ll 16171781140 16149344101 -0.14%
  darktable/DngOpcodes.cpp.ll 2123576004 2120732832 -0.13%
  hermes/Passes.cpp.ll 1997252275 1994672931 -0.13%
Top 5 regressions:
  rust-analyzer-rs/4dj9fscdf5c509wz.ll 145086655 148887948 +2.62%
  typst-rs/p1dgiootfedk7bo.ll 488966323 501623041 +2.59%
  rust-analyzer-rs/1bjrygtvfxna7kin.ll 5561255586 5677108765 +2.08%
  actix-rs/125u7gvq3v36sg70.ll 98217713 99929856 +1.74%
  wasmtime-rs/3r0osxvwe4cd326n.ll 9141682789 9286492070 +1.58%

Overall: -0.00945255%
```

fhahn committed:
[EarlyCSE] Rematerialize alignment assumption.

nikic commented on llvm/lib/Transforms/Scalar/EarlyCSE.cpp:
```diff
@@ -1588,6 +1589,18 @@ bool EarlyCSE::processNode(DomTreeNode *Node) {
+        // If the load has align metadata, preserve it via an alignment
+        // assumption. Note that this doesn't use salavageKnowledge, as we need
+        // to create the assumption for the value we replaced the load with.
+        if (auto *AlignMD = Inst.getMetadata(LLVMContext::MD_align)) {
```
You also need `!noundef` for this transform. `!align` by itself only returns poison, while the assumption converts it into IUB.

fhahn replied:
Should be updated, thanks!

fhahn committed:
!fixup check noundef

dtcxzyw commented:
> Unfortunately the diff is too large to display

Fixed. See https://github.com/dtcxzyw/llvm-opt-benchmark/pull/1345.

nikic commented:
I'm still not convinced that doing something like this is a good idea. The core problem is that you can't tell whether the `!align` metadata is sufficiently valuable or not. In Rust, most pointer loads will have `!align` metadata. It's added because adding it is free and sometimes helps, but at the same time, nobody will particularly care if it gets dropped. Doing this conversion means we now have to make a tradeoff between adding the metadata, and the compile-time and optimization impact it will have.

I looked through the IR diffs, and there were plenty of minor optimization regressions and very few improvements. I think the two most common ones are:

 * Dead block not eliminated due to assume. (This is something I tried to fix in the past and it got reverted due to optimization regressions...)
 * Hoisting prevented because one block has an assume at the top and the other doesn't.

I also noticed that in some cases the same assume is repeated many times in the same block -- we should be optimizing those away.

Another interesting data point is that nearly all assumptions were redundant in the sense that they were directly followed by a load or gep+load with the same alignment. (Of course, most of our analysis cannot really use alignment on loads to reason about the alignment of the pointer at some other point.)

fhahn commented:
> I'm still not convinced that doing something like this is a good idea. The core problem is that you can't tell whether the `!align` metadata is sufficiently valuable or not. In Rust, most pointer loads will have `!align` metadata. It's added because adding it is free and sometimes helps, but at the same time, nobody will particularly care if it gets dropped. Doing this conversion means we now have to make a tradeoff between adding the metadata, and the compile-time and optimization impact it will have.
> 

Yeah, when originally working on https://github.com/llvm/llvm-project/pull/108958, I didn't expect much fallout as !align/alignment assumptions isn't super common in C/C++, forgetting about Rust.

> I looked through the IR diffs, and there were plenty of minor optimization regressions and very few improvements. I think the two most common ones are:
> 
> * Dead block not eliminated due to assume. (This is something I tried to fix in the past and it got reverted due to optimization regressions...)
> * Hoisting prevented because one block has an assume at the top and the other doesn't.
> 

Yep, unfortunately there's a number of improvements that are still needed to avoid negative fallout from using assumptions...

> I also noticed that in some cases the same assume is repeated many times in the same block -- we should be optimizing those away.
> 
> Another interesting data point is that nearly all assumptions were redundant in the sense that they were directly followed by a load or gep+load with the same alignment. (Of course, most of our analysis cannot really use alignment on loads to reason about the alignment of the pointer at some other point.)

We might be able to skip rematerializing in those cases by doing some extra analysis. 

I can look into addressing some of the issues mentioned above. 

Overall my main goal is to remove the blockers for https://github.com/llvm/llvm-project/pull/108958 :)

fhahn committed:
Use computeKnownBits.

fhahn committed:
step

fhahn committed:
[PhaseOrdering] Add eest making sure alignment doesn't get dropped.

fhahn committed:
!fixup update tests
